lib/sles4sap_publiccloud.pm

```diff
@@ -216,12 +229,11 @@ sub sles4sap_cleanup {
                 "Cleanup $command FAILED",
                 result => 'fail'
             ) if $_ == 3 && $cleanup_cmd_rc[0];
-            return 0 if $_ == 3 && $cleanup_cmd_rc[0];
+            $ret = 1 if $_ == 3 && $cleanup_cmd_rc[0];
         }
     }
-    record_info('Cleanup finished');
-    return 1;
-
+    record_info('Cleanup finished', "ret:$ret");
```

Comment:
Hmm. So we can have a `@cmd_list = ('ansible', 'terraform')` at most, right? What do you think about recording here the status for both commands?

You could do:
```
....
use List::Util qw(sum);
....
# Do not run destroy if already executed
return 0 if ($args{cleanup_called});    # instead of return $ret
....
my %rets = map { $_ => 0 } @cmd_list();    # instead of my $ret = 0;
for my $command (@cmd_list) {
....
     $rets{$command} = 1 if $_ == 3 && $cleanup_cmd_rc[0];
....
record_info('Cleanup finished', 'ret:' . join(':', %rets));
return sum values %rets;
```

Reply:
> Hmm. So we can have a `@cmd_list = ('ansible', 'terraform')` at most, right? 

Yes, it is like mr_test only have terraform, all other tests have both. Nothing more than two. For each of this two we have the retry mechanism in place.

> You could do:
> 
> ```
> # Do not run destroy if already executed
> return 0 if ($args{cleanup_called});    # instead of return $ret
> ```

I'm fine with it
> ```
> my %rets = map { $_ => 0 } @cmd_list();    # instead of my $ret = 0;
> ```

Is it so now `{terraform => 0 , ansible => 0}` ?

> ```
> for my $command (@cmd_list) {
> ....
>      $rets{$command} = 1 if $_ == 3 && $cleanup_cmd_rc[0];
> ....
> ```

Is it here recording result per command right? If so I'm fine with it.

> ```
> record_info('Cleanup finished', 'ret:' . join(':', %rets));
> return sum values %rets;
> ```
 
And keep returning a single integer by summing all the ret across all the commands. Again if it is like, so I'm fine as the return does not change.

All in all what we gain from this proposal? Number of retry is the same, behavior in case of failure is the same (terraform is always executed). Do we gain a more expressive return code? Like not a flat 0/1 but 0 in case of PASS and ret>0 and equal to the number of retry in case of failure? What is, and what will be the return if terraform destroy pass at the second attempt? It should be 0.

Reply:
> Is it so now `{terraform => 0 , ansible => 0}` ?

Yes. Creates a hash where the keys are the elements in `@cmd_list` and initializes all to 0.

> Is it here recording result per command right? If so I'm fine with it.

Yes.

> All in all what we gain from this proposal?

We have a text like: `ret:ansible:1:terraform:0` at the end of the call instead of `ret:1`. If we see a 1 in either of them, we know that was the one that failed. It would help in review IMO.